parsers/ES.py

```diff
@@ -30,91 +27,113 @@
-ZONE_KEYS = Literal[
-    "ES",  # Spain
-    "ES-CE",  # Ceuta
-    "ES-CN-FVLZ",  # Fuerteventura/Lanzarote
-    "ES-CN-GC",  # Gran Canaria
-    "ES-CN-HI",  # El Hierro
-    "ES-CN-IG",  # Isla de la Gomera
-    "ES-CN-LP",  # La Palma
-    "ES-CN-TE",  # Tenerife
-    "ES-IB",  # Balearic Islands
-    "ES-IB-FO",  # Formentera
-    "ES-IB-IZ",  # Ibiza
-    "ES-IB-MA",  # Mallorca
-    "ES-IB-ME",  # Menorca
-    "ES-ML",  # Melilla
-]
-
-# TODO: Update floors to be non zero.
-# Minimum valid zone demand. This is used to eliminate some cases
-# where generation for one or more modes is obviously missing.
-ZONE_FLOORS: Dict[ZONE_KEYS, float] = {
-    "ES": 0,
-    "ES-CE": 0,
-    "ES-CN-FVLZ": 50,
-    "ES-CN-GC": 150,
-    "ES-CN-HI": 2,
-    "ES-CN-IG": 3,
-    "ES-CN-LP": 10,
-    "ES-CN-TE": 150,
-    "ES-IB": 0,
-    "ES-IB-FO": 0,
-    "ES-IB-IZ": 0,
-    "ES-IB-MA": 0,
-    "ES-IB-ME": 0,
-    "ES-ML": 0,
-}
 
-ZONE_FUNCTION_MAP: Dict[ZONE_KEYS, Callable] = {
-    "ES": IberianPeninsula,
-    "ES-CE": Ceuta,
-    "ES-CN-FVLZ": LanzaroteFuerteventura,
-    "ES-CN-GC": GranCanaria,
-    "ES-CN-HI": ElHierro,
-    "ES-CN-IG": Gomera,
-    "ES-CN-LP": LaPalma,
-    "ES-CN-TE": Tenerife,
-    "ES-IB": BalearicIslands,
-    "ES-IB-FO": Formentera,
-    "ES-IB-IZ": Ibiza,
-    "ES-IB-MA": Mallorca,
-    "ES-IB-ME": Menorca,
-    "ES-ML": Melilla,
+ZONE_FUNCTION_MAP: Dict[ZoneKey, Callable] = {
+    ZoneKey("ES"): IberianPeninsula,
+    ZoneKey("ES-CE"): Ceuta,
+    ZoneKey("ES-CN-FVLZ"): LanzaroteFuerteventura,
+    ZoneKey("ES-CN-GC"): GranCanaria,
+    ZoneKey("ES-CN-HI"): ElHierro,
+    ZoneKey("ES-CN-IG"): Gomera,
+    ZoneKey("ES-CN-LP"): LaPalma,
+    ZoneKey("ES-CN-TE"): Tenerife,
+    ZoneKey("ES-IB-FO"): Formentera,
+    ZoneKey("ES-IB-IZ"): Ibiza,
+    ZoneKey("ES-IB-MA"): Mallorca,
+    ZoneKey("ES-IB-ME"): Menorca,
+    ZoneKey("ES-ML"): Melilla,
 }
 
 EXCHANGE_FUNCTION_MAP: Dict[str, Callable] = {
-    "ES->ES-IB": BalearicIslands,
     "ES->ES-IB-MA": Mallorca,
     "ES-IB-IZ->ES-IB-MA": Mallorca,
     "ES-IB-FO->ES-IB-IZ": Formentera,
     "ES-IB-MA->ES-IB-ME": Mallorca,
 }
 
+# dict of REE production types to Electricity Maps production types
+PRODUCTION_MAPPING = {
+    "carbon": "coal",
+    "combined": "gas",
+    "diesel": "oil",
+    "gas": "gas",
+    "hydraulic": "hydro",
+    "nuclear": "nuclear",
+    "other": "unknown",
+    "solar": "solar",
+    "vapor": "oil",
+    "waste": "biomass",
+    "wind": "wind",
+}
 
-def fetch_island_data(
-    zone_key: ZONE_KEYS, session: Session, target_datetime: Optional[datetime]
-) -> List[Response]:
-    if isinstance(target_datetime, datetime):
-        date = target_datetime.strftime("%Y-%m-%d")
+
+def check_valid_parameters(
+    zone_key: ZoneKey,
+    session: Optional[Session],
+    target_datetime: Optional[datetime],
+    logger: Logger,
+    exchange: bool = False,
+):
+    """Raise an exception if the parameters are not valid for this parser."""
+    if exchange:
```

Comment:
Instead of using a boolean attribute, I think it would be better to use the supplied zone_key and check if it is in the exchange config.
This way we avoid making mistakes of saying this is an exchange while it is not.

Reply:
Good suggestion, I'll change it.